src/server/pfs/server/driver.go

```diff
@@ -1854,23 +1858,64 @@ func (d *driver) putFile(ctx context.Context, file *pfs.File, delimiter pfs.Deli
 	}
 	if commitInfo != nil && commitInfo.Finished != nil {
 		if branch == "" {
-			return pfsserver.ErrCommitFinished{file.Commit}
+			return pfsserver.ErrCommitFinished{commit}
 		}
 		oneOff = true
 	}
 
+	var files []*pfs.File
+	var putFilePaths []string
+	var putFileRecords []*pfs.PutFileRecords
+	var mu sync.Mutex
+	if err := forEachPutFile(s, func(req *pfs.PutFileRequest, r io.Reader) error {
+		records, err := d.putFile(ctx, req.File, req.Delimiter, req.TargetFileDatums, req.TargetFileBytes, req.OverwriteIndex, r)
+		if err != nil {
+			return err
+		}
+		mu.Lock()
+		defer mu.Unlock()
+		files = append(files, req.File)
```

Comment:
Is there not a way to do this without having to keep track of info for every file we have seen? As far as the upsert is concerned, it seems like it might be as simple as calling the upsert after each putfile, but the one off stuff might be a bit more tricky.

Reply:
There is, and I considered doing it but it would require making some fairly big changes to how `makeCommit` works. I figured that would make more sense as a separate pr rather than bundling it into this one. I also think it likely makes sense to wait until your performance changes are merged in because how we stream stuff in is probably going to be affected in some ways by your hashtree changes.